// @hyperlane-xyz/starknet-core/dist/artifacts/index.js
import { contracts_ERC4626Mock as contracts_ERC4626Mock_sierra } from "./contracts_ERC4626Mock.contract_class.js";
import { contracts_ERC4626YieldSharingMock as contracts_ERC4626YieldSharingMock_sierra } from "./contracts_ERC4626YieldSharingMock.contract_class.js";
import { contracts_EnumerableMapHolder as contracts_EnumerableMapHolder_sierra } from "./contracts_EnumerableMapHolder.contract_class.js";
import { contracts_Ether as contracts_Ether_sierra } from "./contracts_Ether.contract_class.js";
import { contracts_MockAccount as contracts_MockAccount_sierra } from "./contracts_MockAccount.contract_class.js";
import { contracts_MockHypERC721URIStorage as contracts_MockHypERC721URIStorage_sierra } from "./contracts_MockHypERC721URIStorage.contract_class.js";
import { contracts_MockMailbox as contracts_MockMailbox_sierra } from "./contracts_MockMailbox.contract_class.js";
import { contracts_MockParadexDex as contracts_MockParadexDex_sierra } from "./contracts_MockParadexDex.contract_class.js";
import { contracts_MockRateLimited as contracts_MockRateLimited_sierra } from "./contracts_MockRateLimited.contract_class.js";
import { contracts_TestERC20 as contracts_TestERC20_sierra } from "./contracts_TestERC20.contract_class.js";
import { contracts_TestERC721 as contracts_TestERC721_sierra } from "./contracts_TestERC721.contract_class.js";
import { contracts_TestISM as contracts_TestISM_sierra } from "./contracts_TestISM.contract_class.js";
import { contracts_TestInterchainGasPayment as contracts_TestInterchainGasPayment_sierra } from "./contracts_TestInterchainGasPayment.contract_class.js";
import { contracts_TestPostDispatchHook as contracts_TestPostDispatchHook_sierra } from "./contracts_TestPostDispatchHook.contract_class.js";
import { contracts_XERC20LockboxTest as contracts_XERC20LockboxTest_sierra } from "./contracts_XERC20LockboxTest.contract_class.js";
import { contracts_XERC20Test as contracts_XERC20Test_sierra } from "./contracts_XERC20Test.contract_class.js";
import { contracts_aggregation as contracts_aggregation_sierra } from "./contracts_aggregation.contract_class.js";
import { contracts_default_fallback_routing_ism as contracts_default_fallback_routing_ism_sierra } from "./contracts_default_fallback_routing_ism.contract_class.js";
import { contracts_domain_routing_hook as contracts_domain_routing_hook_sierra } from "./contracts_domain_routing_hook.contract_class.js";
import { contracts_domain_routing_ism as contracts_domain_routing_ism_sierra } from "./contracts_domain_routing_ism.contract_class.js";
import { contracts_fee_hook as contracts_fee_hook_sierra } from "./contracts_fee_hook.contract_class.js";
import { contracts_hook as contracts_hook_sierra } from "./contracts_hook.contract_class.js";
import { contracts_ism as contracts_ism_sierra } from "./contracts_ism.contract_class.js";
import { contracts_mailbox as contracts_mailbox_sierra } from "./contracts_mailbox.contract_class.js";
import { contracts_mailboxClientProxy as contracts_mailboxClientProxy_sierra } from "./contracts_mailboxClientProxy.contract_class.js";
import { contracts_merkle_tree_hook as contracts_merkle_tree_hook_sierra } from "./contracts_merkle_tree_hook.contract_class.js";
import { contracts_merkleroot_multisig_ism as contracts_merkleroot_multisig_ism_sierra } from "./contracts_merkleroot_multisig_ism.contract_class.js";
import { contracts_message_recipient as contracts_message_recipient_sierra } from "./contracts_message_recipient.contract_class.js";
import { contracts_messageid_multisig_ism as contracts_messageid_multisig_ism_sierra } from "./contracts_messageid_multisig_ism.contract_class.js";
import { contracts_mock_fee_token as contracts_mock_fee_token_sierra } from "./contracts_mock_fee_token.contract_class.js";
import { contracts_mock_validator_announce as contracts_mock_validator_announce_sierra } from "./contracts_mock_validator_announce.contract_class.js";
import { contracts_noop_ism as contracts_noop_ism_sierra } from "./contracts_noop_ism.contract_class.js";
import { contracts_pausable_ism as contracts_pausable_ism_sierra } from "./contracts_pausable_ism.contract_class.js";
import { contracts_protocol_fee as contracts_protocol_fee_sierra } from "./contracts_protocol_fee.contract_class.js";
import { contracts_trusted_relayer_ism as contracts_trusted_relayer_ism_sierra } from "./contracts_trusted_relayer_ism.contract_class.js";
import { contracts_validator_announce as contracts_validator_announce_sierra } from "./contracts_validator_announce.contract_class.js";
import { mocks_ERC4626Mock as mocks_ERC4626Mock_sierra } from "./mocks_ERC4626Mock.contract_class.js";
import { mocks_ERC4626YieldSharingMock as mocks_ERC4626YieldSharingMock_sierra } from "./mocks_ERC4626YieldSharingMock.contract_class.js";
import { mocks_EnumerableMapHolder as mocks_EnumerableMapHolder_sierra } from "./mocks_EnumerableMapHolder.contract_class.js";
import { mocks_Ether as mocks_Ether_sierra } from "./mocks_Ether.contract_class.js";
import { mocks_MockAccount as mocks_MockAccount_sierra } from "./mocks_MockAccount.contract_class.js";
import { mocks_MockHypERC721URIStorage as mocks_MockHypERC721URIStorage_sierra } from "./mocks_MockHypERC721URIStorage.contract_class.js";
import { mocks_MockMailbox as mocks_MockMailbox_sierra } from "./mocks_MockMailbox.contract_class.js";
import { mocks_MockParadexDex as mocks_MockParadexDex_sierra } from "./mocks_MockParadexDex.contract_class.js";
import { mocks_MockRateLimited as mocks_MockRateLimited_sierra } from "./mocks_MockRateLimited.contract_class.js";
import { mocks_TestERC20 as mocks_TestERC20_sierra } from "./mocks_TestERC20.contract_class.js";
import { mocks_TestERC721 as mocks_TestERC721_sierra } from "./mocks_TestERC721.contract_class.js";
import { mocks_TestISM as mocks_TestISM_sierra } from "./mocks_TestISM.contract_class.js";
import { mocks_TestInterchainGasPayment as mocks_TestInterchainGasPayment_sierra } from "./mocks_TestInterchainGasPayment.contract_class.js";
import { mocks_TestPostDispatchHook as mocks_TestPostDispatchHook_sierra } from "./mocks_TestPostDispatchHook.contract_class.js";
import { mocks_XERC20LockboxTest as mocks_XERC20LockboxTest_sierra } from "./mocks_XERC20LockboxTest.contract_class.js";
import { mocks_XERC20Test as mocks_XERC20Test_sierra } from "./mocks_XERC20Test.contract_class.js";
import { mocks_fee_hook as mocks_fee_hook_sierra } from "./mocks_fee_hook.contract_class.js";
import { mocks_hook as mocks_hook_sierra } from "./mocks_hook.contract_class.js";
import { mocks_ism as mocks_ism_sierra } from "./mocks_ism.contract_class.js";
import { mocks_message_recipient as mocks_message_recipient_sierra } from "./mocks_message_recipient.contract_class.js";
import { mocks_mock_fee_token as mocks_mock_fee_token_sierra } from "./mocks_mock_fee_token.contract_class.js";
import { mocks_mock_validator_announce as mocks_mock_validator_announce_sierra } from "./mocks_mock_validator_announce.contract_class.js";
import { token_ERC4626Mock as token_ERC4626Mock_sierra } from "./token_ERC4626Mock.contract_class.js";
import { token_ERC4626YieldSharingMock as token_ERC4626YieldSharingMock_sierra } from "./token_ERC4626YieldSharingMock.contract_class.js";
import { token_EnumerableMapHolder as token_EnumerableMapHolder_sierra } from "./token_EnumerableMapHolder.contract_class.js";
import { token_Ether as token_Ether_sierra } from "./token_Ether.contract_class.js";
import { token_FastHypERC20 as token_FastHypERC20_sierra } from "./token_FastHypERC20.contract_class.js";
import { token_FastHypERC20Collateral as token_FastHypERC20Collateral_sierra } from "./token_FastHypERC20Collateral.contract_class.js";
import { token_HypERC20CollateralVaultDeposit as token_HypERC20CollateralVaultDeposit_sierra } from "./token_HypERC20CollateralVaultDeposit.contract_class.js";
import { token_HypERC721URICollateral as token_HypERC721URICollateral_sierra } from "./token_HypERC721URICollateral.contract_class.js";
import { token_HypERC721URIStorage as token_HypERC721URIStorage_sierra } from "./token_HypERC721URIStorage.contract_class.js";
import { token_HypErc20 as token_HypErc20_sierra } from "./token_HypErc20.contract_class.js";
import { token_HypErc20Collateral as token_HypErc20Collateral_sierra } from "./token_HypErc20Collateral.contract_class.js";
import { token_HypErc20DexCollateral as token_HypErc20DexCollateral_sierra } from "./token_HypErc20DexCollateral.contract_class.js";
import { token_HypErc20Vault as token_HypErc20Vault_sierra } from "./token_HypErc20Vault.contract_class.js";
import { token_HypErc20VaultCollateral as token_HypErc20VaultCollateral_sierra } from "./token_HypErc20VaultCollateral.contract_class.js";
import { token_HypErc721 as token_HypErc721_sierra } from "./token_HypErc721.contract_class.js";
import { token_HypErc721Collateral as token_HypErc721Collateral_sierra } from "./token_HypErc721Collateral.contract_class.js";
import { token_HypFiatToken as token_HypFiatToken_sierra } from "./token_HypFiatToken.contract_class.js";
import { token_HypNative as token_HypNative_sierra } from "./token_HypNative.contract_class.js";
import { token_HypNativeScaled as token_HypNativeScaled_sierra } from "./token_HypNativeScaled.contract_class.js";
import { token_HypXERC20 as token_HypXERC20_sierra } from "./token_HypXERC20.contract_class.js";
import { token_HypXERC20Lockbox as token_HypXERC20Lockbox_sierra } from "./token_HypXERC20Lockbox.contract_class.js";
import { token_MockAccount as token_MockAccount_sierra } from "./token_MockAccount.contract_class.js";
import { token_MockHypERC721URIStorage as token_MockHypERC721URIStorage_sierra } from "./token_MockHypERC721URIStorage.contract_class.js";
import { token_MockMailbox as token_MockMailbox_sierra } from "./token_MockMailbox.contract_class.js";
import { token_MockParadexDex as token_MockParadexDex_sierra } from "./token_MockParadexDex.contract_class.js";
import { token_MockRateLimited as token_MockRateLimited_sierra } from "./token_MockRateLimited.contract_class.js";
import { token_TestERC20 as token_TestERC20_sierra } from "./token_TestERC20.contract_class.js";
import { token_TestERC721 as token_TestERC721_sierra } from "./token_TestERC721.contract_class.js";
import { token_TestISM as token_TestISM_sierra } from "./token_TestISM.contract_class.js";
import { token_TestInterchainGasPayment as token_TestInterchainGasPayment_sierra } from "./token_TestInterchainGasPayment.contract_class.js";
import { token_TestPostDispatchHook as token_TestPostDispatchHook_sierra } from "./token_TestPostDispatchHook.contract_class.js";
import { token_XERC20LockboxTest as token_XERC20LockboxTest_sierra } from "./token_XERC20LockboxTest.contract_class.js";
import { token_XERC20Test as token_XERC20Test_sierra } from "./token_XERC20Test.contract_class.js";
import { token_fee_hook as token_fee_hook_sierra } from "./token_fee_hook.contract_class.js";
import { token_hook as token_hook_sierra } from "./token_hook.contract_class.js";
import { token_ism as token_ism_sierra } from "./token_ism.contract_class.js";
import { token_message_recipient as token_message_recipient_sierra } from "./token_message_recipient.contract_class.js";
import { token_mock_fee_token as token_mock_fee_token_sierra } from "./token_mock_fee_token.contract_class.js";
import { token_mock_validator_announce as token_mock_validator_announce_sierra } from "./token_mock_validator_announce.contract_class.js";

export const starknetContracts = {
  contracts: {
    ERC4626Mock: { contract_class: contracts_ERC4626Mock_sierra },
    ERC4626YieldSharingMock: {
      contract_class: contracts_ERC4626YieldSharingMock_sierra,
    },
    EnumerableMapHolder: {
      contract_class: contracts_EnumerableMapHolder_sierra,
    },
    Ether: { contract_class: contracts_Ether_sierra },
    MockAccount: { contract_class: contracts_MockAccount_sierra },
    MockHypERC721URIStorage: {
      contract_class: contracts_MockHypERC721URIStorage_sierra,
    },
    MockMailbox: { contract_class: contracts_MockMailbox_sierra },
    MockParadexDex: { contract_class: contracts_MockParadexDex_sierra },
    MockRateLimited: { contract_class: contracts_MockRateLimited_sierra },
    TestERC20: { contract_class: contracts_TestERC20_sierra },
    TestERC721: { contract_class: contracts_TestERC721_sierra },
    TestISM: { contract_class: contracts_TestISM_sierra },
    TestInterchainGasPayment: {
      contract_class: contracts_TestInterchainGasPayment_sierra,
    },
    TestPostDispatchHook: {
      contract_class: contracts_TestPostDispatchHook_sierra,
    },
    XERC20LockboxTest: { contract_class: contracts_XERC20LockboxTest_sierra },
    XERC20Test: { contract_class: contracts_XERC20Test_sierra },
    aggregation: { contract_class: contracts_aggregation_sierra },
    default_fallback_routing_ism: {
      contract_class: contracts_default_fallback_routing_ism_sierra,
    },
    domain_routing_hook: {
      contract_class: contracts_domain_routing_hook_sierra,
    },
    domain_routing_ism: { contract_class: contracts_domain_routing_ism_sierra },
    fee_hook: { contract_class: contracts_fee_hook_sierra },
    hook: { contract_class: contracts_hook_sierra },
    ism: { contract_class: contracts_ism_sierra },
    mailbox: { contract_class: contracts_mailbox_sierra },
    mailboxClientProxy: { contract_class: contracts_mailboxClientProxy_sierra },
    merkle_tree_hook: { contract_class: contracts_merkle_tree_hook_sierra },
    merkleroot_multisig_ism: {
      contract_class: contracts_merkleroot_multisig_ism_sierra,
    },
    message_recipient: { contract_class: contracts_message_recipient_sierra },
    messageid_multisig_ism: {
      contract_class: contracts_messageid_multisig_ism_sierra,
    },
    mock_fee_token: { contract_class: contracts_mock_fee_token_sierra },
    mock_validator_announce: {
      contract_class: contracts_mock_validator_announce_sierra,
    },
    noop_ism: { contract_class: contracts_noop_ism_sierra },
    pausable_ism: { contract_class: contracts_pausable_ism_sierra },
    protocol_fee: { contract_class: contracts_protocol_fee_sierra },
    trusted_relayer_ism: {
      contract_class: contracts_trusted_relayer_ism_sierra,
    },
    validator_announce: { contract_class: contracts_validator_announce_sierra },
  },
  token: {
    ERC4626Mock: { contract_class: token_ERC4626Mock_sierra },
    ERC4626YieldSharingMock: {
      contract_class: token_ERC4626YieldSharingMock_sierra,
    },
    EnumerableMapHolder: { contract_class: token_EnumerableMapHolder_sierra },
    Ether: { contract_class: token_Ether_sierra },
    FastHypERC20: { contract_class: token_FastHypERC20_sierra },
    FastHypERC20Collateral: {
      contract_class: token_FastHypERC20Collateral_sierra,
    },
    HypERC20CollateralVaultDeposit: {
      contract_class: token_HypERC20CollateralVaultDeposit_sierra,
    },
    HypERC721URICollateral: {
      contract_class: token_HypERC721URICollateral_sierra,
    },
    HypERC721URIStorage: { contract_class: token_HypERC721URIStorage_sierra },
    HypErc20: { contract_class: token_HypErc20_sierra },
    HypErc20Collateral: { contract_class: token_HypErc20Collateral_sierra },
    HypErc20DexCollateral: {
      contract_class: token_HypErc20DexCollateral_sierra,
    },
    HypErc20Vault: { contract_class: token_HypErc20Vault_sierra },
    HypErc20VaultCollateral: {
      contract_class: token_HypErc20VaultCollateral_sierra,
    },
    HypErc721: { contract_class: token_HypErc721_sierra },
    HypErc721Collateral: { contract_class: token_HypErc721Collateral_sierra },
    HypFiatToken: { contract_class: token_HypFiatToken_sierra },
    HypNative: { contract_class: token_HypNative_sierra },
    HypNativeScaled: { contract_class: token_HypNativeScaled_sierra },
    HypXERC20: { contract_class: token_HypXERC20_sierra },
    HypXERC20Lockbox: { contract_class: token_HypXERC20Lockbox_sierra },
    MockAccount: { contract_class: token_MockAccount_sierra },
    MockHypERC721URIStorage: {
      contract_class: token_MockHypERC721URIStorage_sierra,
    },
    MockMailbox: { contract_class: token_MockMailbox_sierra },
    MockParadexDex: { contract_class: token_MockParadexDex_sierra },
    MockRateLimited: { contract_class: token_MockRateLimited_sierra },
    TestERC20: { contract_class: token_TestERC20_sierra },
    TestERC721: { contract_class: token_TestERC721_sierra },
    TestISM: { contract_class: token_TestISM_sierra },
    TestInterchainGasPayment: {
      contract_class: token_TestInterchainGasPayment_sierra,
    },
    TestPostDispatchHook: { contract_class: token_TestPostDispatchHook_sierra },
    XERC20LockboxTest: { contract_class: token_XERC20LockboxTest_sierra },
    XERC20Test: { contract_class: token_XERC20Test_sierra },
    fee_hook: { contract_class: token_fee_hook_sierra },
    hook: { contract_class: token_hook_sierra },
    ism: { contract_class: token_ism_sierra },
    message_recipient: { contract_class: token_message_recipient_sierra },
    mock_fee_token: { contract_class: token_mock_fee_token_sierra },
    mock_validator_announce: {
      contract_class: token_mock_validator_announce_sierra,
    },
  },
  mocks: {
    ERC4626Mock: { contract_class: mocks_ERC4626Mock_sierra },
    ERC4626YieldSharingMock: {
      contract_class: mocks_ERC4626YieldSharingMock_sierra,
    },
    EnumerableMapHolder: { contract_class: mocks_EnumerableMapHolder_sierra },
    Ether: { contract_class: mocks_Ether_sierra },
    MockAccount: { contract_class: mocks_MockAccount_sierra },
    MockHypERC721URIStorage: {
      contract_class: mocks_MockHypERC721URIStorage_sierra,
    },
    MockMailbox: { contract_class: mocks_MockMailbox_sierra },
    MockParadexDex: { contract_class: mocks_MockParadexDex_sierra },
    MockRateLimited: { contract_class: mocks_MockRateLimited_sierra },
    TestERC20: { contract_class: mocks_TestERC20_sierra },
    TestERC721: { contract_class: mocks_TestERC721_sierra },
    TestISM: { contract_class: mocks_TestISM_sierra },
    TestInterchainGasPayment: {
      contract_class: mocks_TestInterchainGasPayment_sierra,
    },
    TestPostDispatchHook: { contract_class: mocks_TestPostDispatchHook_sierra },
    XERC20LockboxTest: { contract_class: mocks_XERC20LockboxTest_sierra },
    XERC20Test: { contract_class: mocks_XERC20Test_sierra },
    fee_hook: { contract_class: mocks_fee_hook_sierra },
    hook: { contract_class: mocks_hook_sierra },
    ism: { contract_class: mocks_ism_sierra },
    message_recipient: { contract_class: mocks_message_recipient_sierra },
    mock_fee_token: { contract_class: mocks_mock_fee_token_sierra },
    mock_validator_announce: {
      contract_class: mocks_mock_validator_announce_sierra,
    },
  },
};
