// @hyperlane-xyz/core/dist/typechain/factories/contracts/isms/ccip-read/CommitmentReadIsm.sol/CommitmentReadIsmService__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes",
                name: "_message",
                type: "bytes",
            },
        ],
        name: "getCallsFromRevealMessage",
        outputs: [
            {
                internalType: "address",
                name: "ica",
                type: "address",
            },
            {
                internalType: "bytes32",
                name: "salt",
                type: "bytes32",
            },
            {
                components: [
                    {
                        internalType: "bytes32",
                        name: "to",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint256",
                        name: "value",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes",
                        name: "data",
                        type: "bytes",
                    },
                ],
                internalType: "struct CallLib.Call[]",
                name: "_calls",
                type: "tuple[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class CommitmentReadIsmService__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
