// @hyperlane-xyz/core/dist/typechain/factories/contracts/isms/hook/OPL2ToL1CcipReadIsm.sol/OpL2toL1Service__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes",
                name: "_message",
                type: "bytes",
            },
        ],
        name: "getFinalizeWithdrawalTx",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "nonce",
                        type: "uint256",
                    },
                    {
                        internalType: "address",
                        name: "sender",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "value",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "gasLimit",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes",
                        name: "data",
                        type: "bytes",
                    },
                ],
                internalType: "struct IOptimismPortal.WithdrawalTransaction",
                name: "_tx",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "_message",
                type: "bytes",
            },
        ],
        name: "getWithdrawalProof",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "nonce",
                        type: "uint256",
                    },
                    {
                        internalType: "address",
                        name: "sender",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "value",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "gasLimit",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes",
                        name: "data",
                        type: "bytes",
                    },
                ],
                internalType: "struct IOptimismPortal.WithdrawalTransaction",
                name: "_tx",
                type: "tuple",
            },
            {
                internalType: "uint256",
                name: "_disputeGameIndex",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "bytes32",
                        name: "version",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "stateRoot",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "messagePasserStorageRoot",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "latestBlockhash",
                        type: "bytes32",
                    },
                ],
                internalType: "struct IOptimismPortal.OutputRootProof",
                name: "_outputRootProof",
                type: "tuple",
            },
            {
                internalType: "bytes[]",
                name: "_withdrawalProof",
                type: "bytes[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class OpL2toL1Service__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
