// @hyperlane-xyz/core/dist/typechain/factories/contracts/interfaces/optimism/IOptimismPortal__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "nonce",
                        type: "uint256",
                    },
                    {
                        internalType: "address",
                        name: "sender",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "value",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "gasLimit",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes",
                        name: "data",
                        type: "bytes",
                    },
                ],
                internalType: "struct IOptimismPortal.WithdrawalTransaction",
                name: "_tx",
                type: "tuple",
            },
        ],
        name: "finalizeWithdrawalTransaction",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "_withdrawalHash",
                type: "bytes32",
            },
        ],
        name: "finalizedWithdrawals",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "nonce",
                        type: "uint256",
                    },
                    {
                        internalType: "address",
                        name: "sender",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "target",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "value",
                        type: "uint256",
                    },
                    {
                        internalType: "uint256",
                        name: "gasLimit",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes",
                        name: "data",
                        type: "bytes",
                    },
                ],
                internalType: "struct IOptimismPortal.WithdrawalTransaction",
                name: "_tx",
                type: "tuple",
            },
            {
                internalType: "uint256",
                name: "_disputeGameIndex",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "bytes32",
                        name: "version",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "stateRoot",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "messagePasserStorageRoot",
                        type: "bytes32",
                    },
                    {
                        internalType: "bytes32",
                        name: "latestBlockhash",
                        type: "bytes32",
                    },
                ],
                internalType: "struct IOptimismPortal.OutputRootProof",
                name: "_outputRootProof",
                type: "tuple",
            },
            {
                internalType: "bytes[]",
                name: "_withdrawalProof",
                type: "bytes[]",
            },
        ],
        name: "proveWithdrawalTransaction",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "withdrawalHash",
                type: "bytes32",
            },
        ],
        name: "provenWithdrawals",
        outputs: [
            {
                components: [
                    {
                        internalType: "bytes32",
                        name: "outputRoot",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint128",
                        name: "timestamp",
                        type: "uint128",
                    },
                    {
                        internalType: "uint128",
                        name: "l2OutputIndex",
                        type: "uint128",
                    },
                ],
                internalType: "struct IOptimismPortal.ProvenWithdrawal",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class IOptimismPortal__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
