// @hyperlane-xyz/core/dist/typechain/factories/contracts/interfaces/optimism/IOptimismPortal2__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "withdrawalHash",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "msgSender",
                type: "address",
            },
        ],
        name: "provenWithdrawals",
        outputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "disputeGameProxy",
                        type: "address",
                    },
                    {
                        internalType: "uint64",
                        name: "timestamp",
                        type: "uint64",
                    },
                ],
                internalType: "struct IOptimismPortal2.ProvenWithdrawal",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class IOptimismPortal2__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
