// @hyperlane-xyz/core/dist/typechain/factories/contracts/interfaces/cctp/ITokenMessenger__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "uint64",
                name: "nonce",
                type: "uint64",
            },
            {
                indexed: true,
                internalType: "address",
                name: "burnToken",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                indexed: true,
                internalType: "address",
                name: "depositor",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "mintRecipient",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint32",
                name: "destinationDomain",
                type: "uint32",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "destinationTokenMessenger",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "destinationCaller",
                type: "bytes32",
            },
        ],
        name: "DepositForBurn",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "amount",
                type: "uint256",
            },
            {
                internalType: "uint32",
                name: "destinationDomain",
                type: "uint32",
            },
            {
                internalType: "bytes32",
                name: "mintRecipient",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "burnToken",
                type: "address",
            },
        ],
        name: "depositForBurn",
        outputs: [
            {
                internalType: "uint64",
                name: "_nonce",
                type: "uint64",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "messageBodyVersion",
        outputs: [
            {
                internalType: "uint32",
                name: "",
                type: "uint32",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class ITokenMessenger__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
