// @hyperlane-xyz/core/dist/typechain/factories/contracts/interfaces/cctp/IMessageTransmitter.sol/IRelayer__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes",
                name: "originalMessage",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "originalAttestation",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "newMessageBody",
                type: "bytes",
            },
            {
                internalType: "bytes32",
                name: "newDestinationCaller",
                type: "bytes32",
            },
        ],
        name: "replaceMessage",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint32",
                name: "destinationDomain",
                type: "uint32",
            },
            {
                internalType: "bytes32",
                name: "recipient",
                type: "bytes32",
            },
            {
                internalType: "bytes",
                name: "messageBody",
                type: "bytes",
            },
        ],
        name: "sendMessage",
        outputs: [
            {
                internalType: "uint64",
                name: "",
                type: "uint64",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint32",
                name: "destinationDomain",
                type: "uint32",
            },
            {
                internalType: "bytes32",
                name: "recipient",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "destinationCaller",
                type: "bytes32",
            },
            {
                internalType: "bytes",
                name: "messageBody",
                type: "bytes",
            },
        ],
        name: "sendMessageWithCaller",
        outputs: [
            {
                internalType: "uint64",
                name: "",
                type: "uint64",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class IRelayer__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
