// @hyperlane-xyz/sdk/dist/token/config.js
export var TokenType;
(function (TokenType) {
    TokenType["synthetic"] = "synthetic";
    TokenType["syntheticRebase"] = "syntheticRebase";
    TokenType["syntheticUri"] = "syntheticUri";
    TokenType["collateral"] = "collateral";
    TokenType["collateralVault"] = "collateralVault";
    TokenType["collateralVaultRebase"] = "collateralVaultRebase";
    TokenType["XERC20"] = "xERC20";
    TokenType["XERC20Lockbox"] = "xERC20Lockbox";
    TokenType["collateralFiat"] = "collateralFiat";
    TokenType["collateralUri"] = "collateralUri";
    TokenType["collateralCctp"] = "collateralCctp";
    TokenType["native"] = "native";
    TokenType["nativeOpL2"] = "nativeOpL2";
    TokenType["nativeOpL1"] = "nativeOpL1";
    // backwards compatible alias to native
    TokenType["nativeScaled"] = "nativeScaled";
})(TokenType || (TokenType = {}));
// A token is defined movable collateral if its solidity contract implementation
// is a subclass of MovableCollateralRouter
const isMovableCollateralTokenTypeMap = {
    [TokenType.XERC20]: false,
    [TokenType.XERC20Lockbox]: false,
    [TokenType.collateral]: true,
    [TokenType.collateralCctp]: false,
    [TokenType.collateralFiat]: false,
    [TokenType.collateralUri]: false,
    [TokenType.collateralVault]: true,
    [TokenType.collateralVaultRebase]: true,
    [TokenType.native]: true,
    [TokenType.nativeOpL1]: false,
    [TokenType.nativeOpL2]: false,
    [TokenType.nativeScaled]: true,
    [TokenType.synthetic]: false,
    [TokenType.syntheticRebase]: false,
    [TokenType.syntheticUri]: false,
};
export function isMovableCollateralTokenType(type) {
    return !!isMovableCollateralTokenTypeMap[type];
}
export const gasOverhead = (tokenType) => {
    switch (tokenType) {
        case TokenType.synthetic:
            return 64000;
        case TokenType.native:
            return 44000;
        default:
            return 68000;
    }
};
export const NON_ZERO_SENDER_ADDRESS = '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266';
